### PR TITLE
Replace the widget-based table viewer with a simpler pure-HTML table

### LINKED ISSUE #2577: Sort columns alphabetically in `scipp.table()`
Note that this is already done in `_repr_html_`, so we should be consistent.

### PR DESCRIPTION
**Pull request by nvaytet**
This replaces the old implementation for `sc.table` that used widgets with a simpler pure-html solution.

This only handles 1D data, and no longer 0D (as the old one used to).
If the data entries have 0D attributes, they are dropped.
0D masks are broadcast to the size of the data.

There is no longer a slider to scroll large tables.

Fixes #2577

### CODE REVIEW

File: src/scipp/html/table.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    if isinstance(obj, DataArray):
+        return Dataset({obj.name: obj})
+    if isinstance(obj, Variable):
+        return Dataset(data={"": obj})
+    if isinstance(obj, dict):
+        return Dataset(obj)
+    return obj
+
+
+def table(obj: Union[VariableLike, dict], max_rows: Optional[int] = 20):
```
Comment (jl-wynen) [inline src/scipp/html/table.py:173]:
```suggestion
def table(obj: Union[VariableLike, Dict[str, Union[Variable, DataArray]]], max_rows: int = 20):
```

File: src/scipp/html/table.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+    Possible inputs are:
+     - Variable
+     - DataArray
+     - Dataset
+     - dict of Variable
+     - dict of DataArray
+
+    Inputs must be one-dimensional. Zero-dimensional data members, attributes and
+    coordinates are stripped. Zero-dimensional masks are broadcast.
```
Comment (jl-wynen) [inline src/scipp/html/table.py:184]: Why treat masks and coords/attrs differently?

File: src/scipp/html/table.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+import numpy as np
+from typing import List, Optional, Union
+from .. import DataArray, Dataset, Variable
+from ..typing import VariableLike
+from .. import DType
+
+CENTER = 'style="text-align: center;"'
+WITH_BORDER = 'style="border-left:1px solid #a9a9a9;"'
+CENTER_BORDER = 'style="text-align: center; border-left:1px solid #a9a9a9;"'
```
Comment (jl-wynen) [inline src/scipp/html/table.py:12]: Looks good. But I have some comments about the style:

- Tables of a single variable have a black border on the left because they are treated like datasets without coords. This looks a bit odd.
- The separators between dataset items don't extend to the row of dataset names. Is this intentional?
- It might help to put a line between the header and the actual data (like pandas).

File: src/scipp/html/table.py (hunk 1 above)
Comment (nvaytet) [reply src/scipp/html/table.py:173]: :+1: for the Dict, why remove the `Optional`? Should I remove `Optional` in other places as well?

File: src/scipp/html/table.py (hunk 2 above)
Comment (nvaytet) [reply src/scipp/html/table.py:184]: So would just just broadcast everything?

File: src/scipp/html/table.py (hunk 1 above)
Comment (jl-wynen) [reply src/scipp/html/table.py:173]: `Optional[T]` is basically an alias for `Union[T, None]`. So if `table` does not support `max_rows=None`, it can't be `Optional`. (This has nothing to do with whether there is a default value.)

File: src/scipp/html/table.py (hunk 2 above)
Comment (jl-wynen) [reply src/scipp/html/table.py:184]: No, this would be noisy when there are many scalar attrs. But why did you broadcast the mask? It means that either all or no elements are masked. Could this be represented more neatly?

File: src/scipp/html/table.py (hunk 1 above)
Comment (nvaytet) [reply src/scipp/html/table.py:173]: > This has nothing to do with whether there is a default value.

Ok I wasn't aware of that subtlety.

File: src/scipp/html/table.py (hunk 2 above)
Comment (nvaytet) [reply src/scipp/html/table.py:184]: Because I think the mask information is still important. If there is a 0D mask, it will apply to everything. Just like when you plot things, you need to broadcast the masks.

File: src/scipp/html/table.py (hunk 3 above)
Comment (nvaytet) [reply src/scipp/html/table.py:12]: > Tables of a single variable have a black border on the left because they are treated like datasets without coords. This looks a bit odd.

Yes, it was the simplest thing to leave it as is. But I can try to fix if you want.

> The separators between dataset items don't extend to the row of dataset names. Is this intentional?

Yes, I thought it looked cleaner without the lines going all the way at the top. But we can change.
Screenshot at 2022-05-30 14-15-55

> It might help to put a line between the header and the actual data (like pandas).

:+1:

Comment (jl-wynen) [reply src/scipp/html/table.py:12]: > Yes, it was the simplest thing to leave it as is. But I can try to fix if you want.

Depends on how much extra code this requires. It should be rare to use `table` for variables...

> Yes, I thought it looked cleaner without the lines going all the way at the top. But we can change.

I find both a bit strange because the names are so short. But I prefer the left because it makes it clearer that 'a' and 'b' are part of the table.

File: src/scipp/html/table.py (hunk 2 above)
Comment (jl-wynen) [reply src/scipp/html/table.py:184]: It might not be entirely clear that the mask affects all elements when the data is cut off in the table. But I don't see a particularly good way of representing that. Maybe this?
```
Data Masks
[m]   mask
0.1   1
0.2   ...
```

Also, just realised that masks are represented numerically. I think showing them as 'True' / 'False' would be better. (In the off chance that as mask has different values, casting to `bool` should still be fine here?)

Comment (nvaytet) [reply src/scipp/html/table.py:184]: I thought that having as numbers might be nice if someone wants to copy/paste some values into a csv or something.

File: docs/tutorials/introduction.ipynb [hunk 4]
```diff
@@ -161,9 +161,10 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "sc.table(d['alice']) # includes coordinates\n",
-    "sc.table(d['alice'].data) # the variable holding the data, i.e., the dimension labels, units, values, and optional variances\n",
-    "sc.table(d['alice'].values) # just the array of values, shorthand for d['alice'].data.values"
+    "from IPython.display import display\n",
```
Comment (jl-wynen) [inline docs/tutorials/introduction.ipynb:164]: Why did you add calls to `display`

Comment (nvaytet) [reply docs/tutorials/introduction.ipynb:164]: Because I changed the function to return the `HTML()` object, instead of calling `display(HTML())` inside the function and returning None. This was more consistent with the html repr, but I can change it back if we prefer the old behaviour.

Comment (jl-wynen) [reply docs/tutorials/introduction.ipynb:164]: I think the way it works no is better :+1:

**Review by jl-wynen — APPROVED** [review]

File: src/scipp/html/table.py (hunk 2 above)
Comment (SimonHeybrock) [reply src/scipp/html/table.py:188]: I think `True`/`False` is preferable? Not everyone knows how those map to integers?
